tripy/tripy/frontend/cache.py

```diff
@@ -0,0 +1,153 @@
+
+    def _assign_tensor_name(
+        self,
+        tensor: "tripy.frontend.trace.tensor.TraceTensor",
+        tensor_map: Dict[int, str],
+        next_id: List[int],
+        backup_map: Dict[int, str] = None,
+    ) -> str:
+        """
+        Assign or retrieve a tensor name.
+
+        Args:
+            tensor (TraceTensor): The tensor to name.
+            tensor_map (Dict[int, str]): Mapping of tensor ids to names (clean or original).
+            next_id (List[int]): Mutable list for tracking next tensor ID.
+            backup_map (Dict[int, str], optional): Mapping to store original names. Defaults to None.
+
+        Returns:
+            str: The assigned or retrieved tensor name.
+        """
+        t_id = id(tensor)
+
+        # If tensor not in map, assign new name
+        if t_id not in tensor_map:
+            new_name = f"t{next_id[0]}"
+            tensor_map[t_id] = new_name
+            if backup_map is not None:
+                backup_map[t_id] = tensor.name
+            next_id[0] += 1
+
+        return tensor_map[t_id]
+
+    def _update_trace_names(
+        self, trace: "Trace", tensor_map: Dict[int, str], next_id: List[int], backup_map: Dict[int, str] = None
+    ) -> None:
+        """
+        Update names for inputs, outputs, and operations in the trace.
+
+        Args:
+            trace (Trace): The trace to update.
+            tensor_map (Dict[int, str]): Mapping of tensor ids to names.
+            next_id (List[int]): Mutable list for tracking next tensor ID.
+            backup_map (Dict[int, str], optional): Mapping of original tensor names. Defaults to None.
+        """
+        # Update input names
+        for inp in trace.inputs:
+            inp.name = self._assign_tensor_name(inp, tensor_map, next_id, backup_map)
+
+        # Update operation input and output names
+        for op in trace.ops:
+            for inp in op.inputs:
+                inp.name = self._assign_tensor_name(inp, tensor_map, next_id, backup_map)
+            for out in op.outputs:
+                out.name = self._assign_tensor_name(out, tensor_map, next_id, backup_map)
+
+        # Update output names
+        for out in trace.outputs:
+            out.name = self._assign_tensor_name(out, tensor_map, next_id, backup_map)
+
+    def _normalize_trace(self, trace: "Trace") -> str:
+        """
+        Normalize the trace by renaming all tensor names while preserving the structure.
+
+        Args:
+            trace (Trace): The trace to normalize.
+
+        Returns:
+            str: Normalized trace as a string.
+        """
+        # Initialize maps and next tensor ID
+        tensor_map: Dict[int, str] = {}
+        backup_tensor_map: Dict[int, str] = {}
+        next_tensor_id: List[int] = [0]
+
+        # Clean trace names with sequential names
+        self._update_trace_names(trace, tensor_map, next_tensor_id, backup_tensor_map)
+
+        # Get normalized trace string
+        trace_str = str(trace)
```

Comment:
Would be nice to have a context manager that does the renaming like:
```py
with NormalizeTrace(trace) as normalized_trace:
    ...
```
Doesn't have to be done in this PR though.

Reply:
I will make an issue for this and try to it in another MR